File: ios/Classes/AdaptiveCoreFlutterPlugin.swift
import Flutter
import UIKit
import AdaptiveCore

public class AdaptiveCoreFlutterPlugin: NSObject, FlutterPlugin {

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "adaptive_core",
            binaryMessenger: registrar.messenger()
        )
        let instance = AdaptiveCoreFlutterPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [:]
        switch call.method {

        case "initialize":
            guard let clientId = args["clientId"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "clientId is required", details: nil))
                return
            }
            let debug = args["debug"] as? Bool ?? false
            AdaptiveCore.shared.initialize(clientId: clientId, debug: debug)
            result(nil)

        case "login":
            guard let userId    = args["userId"]    as? String,
                  let userName  = args["userName"]  as? String,
                  let userEmail = args["userEmail"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "userId, userName and userEmail are required", details: nil))
                return
            }
            let phoneNumber = args["phoneNumber"] as? String ?? ""
            AdaptiveCore.shared.login(userId: userId, userName: userName, userEmail: userEmail, phoneNumber: phoneNumber)
            result(nil)

        case "logout":
            AdaptiveCore.shared.logout()
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

File: android/src/main/kotlin/com/adaptive/core/AdaptiveCoreFlutterPlugin.kt
package com.adaptive.core

import android.content.Context
import android.os.Handler
import android.os.Looper
import com.adaptive.core.AdaptiveCore
import com.adaptive.core.AdaptiveUser
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch

/** Flutter plugin that bridges the Adaptive Core Android SDK. */
class AdaptiveCoreFlutterPlugin : FlutterPlugin, MethodCallHandler {

    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private val mainHandler = Handler(Looper.getMainLooper())

    // ── FlutterPlugin ─────────────────────────────────────────────────────────

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "adaptive_core")
        channel.setMethodCallHandler(this)
        context = binding.applicationContext
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    // ── MethodCallHandler ─────────────────────────────────────────────────────

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {

            "initialize" -> {
                val clientId = call.argument<String>("clientId")
                    ?: return result.error("INVALID_ARGUMENT", "clientId is required", null)
                val debug = call.argument<Boolean>("debug") ?: false
                try {
                    AdaptiveCore.initialize(context, clientId, debug)
                    result.success(null)
                } catch (e: Exception) {
                    result.error("INITIALIZATION_ERROR", e.message, null)
                }
            }

            "login" -> {
                val userId = call.argument<String>("userId")
                    ?: return result.error("INVALID_ARGUMENT", "userId is required", null)
                val userName = call.argument<String>("userName")
                    ?: return result.error("INVALID_ARGUMENT", "userName is required", null)
                val userEmail = call.argument<String>("userEmail")
                    ?: return result.error("INVALID_ARGUMENT", "userEmail is required", null)
                val phoneNumber = call.argument<String>("phoneNumber")
                    ?: return result.error("INVALID_ARGUMENT", "phoneNumber is required", null)
                try {
                    AdaptiveCore.login(
                        AdaptiveUser(
                            userName = userName,
                            userID = userId,
                            userEmail = userEmail,
                            phoneNumber = phoneNumber,
                        )
                    )
                    result.success(null)
                } catch (e: Exception) {
                    result.error("LOGIN_ERROR", e.message, null)
                }
            }

            "logout" -> {
                try {
                    AdaptiveCore.logout()
                    result.success(null)
                } catch (e: Exception) {
                    result.error("LOGOUT_ERROR", e.message, null)
                }
            }

            else -> result.notImplemented()
        }
    }
}
